### ISSUE TITLE
2.0 regression when searching a dict with functions

### ISSUE THREAD

**Issue by mraspaud** [dpath-maintainers/dpath-python#119]
Hi,

We stumbled upon an issue with the new release (2.0.1) of dpath:

```python

import dpath.util

def foo():
    print('hi')

the_dict = {'fun': foo, 'config': 'something'}

dpath.util.values(the_dict, 'something')
```

crashes with a `TypeError`:
```python-traceback
TypeError: object of type 'function' has no len()
```

This was working fine in dpath <= 1.5.0.
Is this usage not allowed anymore ?

**Comment by sscherfke** [comment]
This also seems to apply to other types:
```
dpath/segments.py", line 16, in kvs
    return zip(range(len(node)), node)
TypeError: object of type 'datetime.datetime' has no len()
```

**Comment by mraspaud** [comment]
@akesterson @calebcase do you have an idea on how to fix this ?

**Comment by larsclaussen** [comment]
happens also with dataclasses as values:
```@dataclass
class Connection:
    group_name: str
    channel_name: str
    last_seen: float
    state: WebsocketConnectionStates

connection = Connection(
    group_name=group_name, 
    channel_name=channel_name, 
    last_seen=time.time(), 
    state=WebsocketConnectionStates.pending
)
d = {"my-key": connection}
from dpath.util import search
search(d, "my*") 

     14         return iter(node.items())
     15     except AttributeError:
---> 16         return zip(range(len(node)), node)
     17 
     18 

TypeError: object of type 'Connection' has no len()



pip list | grep dpath
dpath                         2.0.1 
```

**Comment by mraspaud** [comment]
Looks like there are 3 open PRs now that seem to be fixing this: #120, #122, #123.

**Comment by CD3** [comment]
The error is caused by what `dpath` considers a leaf.

`dpath/segments.py` line 19
``` 
def leaf(thing):                                                                                                     
      '''                                                                                                              
      Return True if thing is a leaf, otherwise False.                                                                 
                                                                                                                       
      leaf(thing) -> bool                                                                                              
      '''                                                                                                              
      leaves = (bytes, str, int, float, bool, type(None))                                                              
                                                                                                                       
      return isinstance(thing, leaves)    

```

The `walk` function (also in `dpath/segments.py`) uses this to determine if it should stop or not.
```
  def walk(obj, location=()):                                                                                          
      '''                                                                                                              
      Yield all valid (segments, value) pairs (from a breadth-first                                                    
      search, right-to-left on sequences).                                                                             
                                                                                                                       
      walk(obj) -> (generator -> (segments, value))                                                                    
      '''                                                                                                              
      if not leaf(obj):       <<<<<<<<<<<<< Right here.                                                                                      
          for k, v in kvs(obj):                                                                                        
              length = None                                                                                            
                                                                                                                       
              try:                                                                                                     
                  length = len(k)                                                                                      
              except:                                                                                                  
                  pass                                                                                                 
                                                                                                                       
              if length is not None and length == 0 and not options.ALLOW_EMPTY_STRING_KEYS:                           
                  raise InvalidKeyName("Empty string keys not allowed without "                                        
                                       "dpath.options.ALLOW_EMPTY_STRING_KEYS=True: "                                  
                                       "{}".format(location + (k,)))                                                   
              yield ((location + (k,)), v)                                                                             
          for k, v in kvs(obj):                                                                                        
              for found in walk(v, location + (k,)):                                                                   
                  yield found      
```
So anything that is not one of the types listes in `leaf` will have `kvs(...)` called on it, which is also in `dpath/segments.py`
```
  def kvs(node):                                                                                                       
      '''                                                                                                              
      Return a (key, value) iterator for the node.                                                                     
                                                                                                                       
      kvs(node) -> (generator -> (key, value))                                                                         
      '''                                                                                                              
      try:                                                                                                             
          return iter(node.items())                                                                                    
      except AttributeError:                                                                                           
          return zip(range(len(node)), node)     <<<<<<<<< Right here
```
If the node doesn't have an `items()` method, then it is passed to `len()`.

This seems like a bug to me. Nodes should be considered leaves if they can't be iterated. Otherewize, there is no way to store user-defined objects in the dict.


The workaround is to replace the `leaf` function (I live python). In my case, my dict only uses `dict` and `list` for branches, so I just did
```
import dpath.utils
dpath.segments.leaf = lambda x: not isinstance(x,(dict,list))
```
If you have other branch types you can list them too.

**Comment by akesterson** [comment]
This is blocked by #136

**Comment by moomoohk** [comment]
Hopefully fixed by merge of #126
